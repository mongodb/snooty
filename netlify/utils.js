import fs from 'fs';
import path from 'path';

/**
 * Parses build hook data from the expected temporary location
 * @returns {object | undefined}
 */
const parseBuildHookData = () => {
  // This file does not currently exist, but should be created on Netlify builds as part of `npm run build:netlify`.
  // The INCOMING_HOOK_BODY env var is not automatically passed along to functions, so we use a txt file to save it
  const relativeFilePath = '../build-hook.txt';
  const buildHookDataString = fs.readFileSync(path.resolve(__dirname, relativeFilePath), 'utf-8');
  if (!buildHookDataString) {
    console.log('No build hook data found.');
    return;
  }
  return JSON.parse(buildHookDataString);
};

export const constructResPayload = (event) => {
  const buildHookData = parseBuildHookData();
  const parsedEventBody = JSON.parse(event.body);
  // This is Netlify's default post-deployment payload. We include it with our custom data in case
  // we want to process any information
  const netlifyPayload = parsedEventBody.payload;
  return {
    netlifyPayload,
    ...buildHookData,
  };
};
