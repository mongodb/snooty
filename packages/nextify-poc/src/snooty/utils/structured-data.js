/**
 * Classes to construct Structured Data JSON.
 * Required props should be read in constructor function (to fail validity).
 * Optional props can be set conditionally.
 * Constant values should be set in the constructor function.
 * Optional overwrites can be set in params as default values
 */

import sanitize from 'sanitize-html';
import { getFullLanguageName } from './get-language';
import { findKeyValuePair } from './find-key-value-pair';
import { getPlaintext } from './get-plaintext';
import { getCompleteBreadcrumbData, getFullBreadcrumbPath } from './get-complete-breadcrumb-data';

// Class name to help Smartling identify all structured data, if needed
export const STRUCTURED_DATA_CLASSNAME = 'structured_data';

export class StructuredData {
  constructor(type) {
    this['@context'] = 'https://schema.org';
    this['@type'] = type;
  }

  isValid() {
    function recursiveValidity(param) {
      // array
      if (Array.isArray(param)) {
        return param.every((e) => recursiveValidity(e));
      }
      // object
      else if (param && typeof param === 'object') {
        return Object.keys(param).every((e) => {
          if (param.hasOwnProperty(e)) return recursiveValidity(param[e]);
          return true;
        });
      }

      // string or number
      return String(param).length > 0;
    }

    return recursiveValidity(this);
  }

  toString() {
    return JSON.stringify(this);
  }

  static addCompanyToName(name) {
    if (!name) {
      return name;
    }
    if (Array.isArray(name)) {
      return name.map(this.addCompanyToName);
    }
    if (name.toLowerCase().includes('mongodb')) {
      return name;
    }
    return `MongoDB ` + name;
  }
}

export class BreadcrumbListSd extends StructuredData {
  constructor({ siteUrl, siteTitle, slug, queriedCrumbs, parentPaths }) {
    super('BreadcrumbList');
    const breadcrumbs = getCompleteBreadcrumbData({ siteUrl, siteTitle, slug, queriedCrumbs, parentPaths });
    this.itemListElement = this.getBreadcrumbList(breadcrumbs, siteUrl);
  }

  /**
   * @param {object[]} breadcrumbs
   * @param {string} siteUrl
   */
  getBreadcrumbList(breadcrumbs, siteUrl) {
    return breadcrumbs.map(({ path, title }, index) => ({
      '@type': 'ListItem',
      position: index + 1,
      name: title,
      item: getFullBreadcrumbPath(siteUrl, path, true),
    }));
  }
}

class HowToSd extends StructuredData {
  constructor({ steps, name }) {
    super('HowTo');

    this.steps = steps;
    this.name = name;
    // image comes from Flora constants
    // https://github.com/10gen/flora/blob/v1.14.2/src/MDBLogo/constants.ts
    this.image =
      'https://webimages.mongodb.com/_com_assets/cms/kuyj2focmkbxv7gh3-stacked_default_slate_blue.svg?auto=format%252Ccompress';
  }
}

export class SoftwareSourceCodeSd extends StructuredData {
  constructor({ code, lang, slug }) {
    super('SoftwareSourceCode');
    this.codeSampleType = 'code snippet';
    // Sanitize all input in case HTML snippets are labeled with different language
    this.text = sanitize(code, { disallowedTagsMode: 'escape' });

    const programmingLanguage = getFullLanguageName(lang, slug);
    if (programmingLanguage) {
      this.programmingLanguage = programmingLanguage;
    }
  }
}

class TechArticleSd extends StructuredData {
  constructor({ headline, mainEntity, genre }) {
    super('TechArticle');

    this.author = {
      '@type': 'Organization',
      name: 'MongoDB Documentation Team',
    };
    this.headline = headline;
    this.mainEntity = mainEntity;
    if (genre) {
      this.genre = genre;
    }
  }
}

export class VideoObjectSd extends StructuredData {
  constructor({ embedUrl, name, uploadDate, thumbnailUrl, description }) {
    super('VideoObject');

    this.embedUrl = embedUrl;
    this.name = name;
    this.uploadDate = uploadDate;
    this.thumbnailUrl = thumbnailUrl;

    if (description) {
      this.description = description;
    }
  }

  isValid() {
    const hasAllReqFields = [this.embedUrl, this.name, this.uploadDate, this.thumbnailUrl].every((val) => !!val);
    return hasAllReqFields && super.isValid();
  }
}

/**
 * get TechArticle Structured Data from page facets and pageTitle.
 * @param   {{category: string, sub_facets: object[]}[]}  facets
 * @param   {string}                                      pageTitle
 * @returns {StructuredData}
 */
export const constructTechArticle = ({ facets, pageTitle }) => {
  // get display name from facets
  function getDisplayName(facet) {
    return facet.display_name;
  }

  // extract genre facets
  function getGenreNames(facets) {
    return facets?.filter((facet) => facet.category === 'genre').map(getDisplayName) || [];
  }

  // extract target product facets
  function getTargetProductsNames(facets) {
    // TODO: these products and sub products need version data from facets
    // https://jira.mongodb.org/browse/DOP-5037
    let res = [];
    const productFacets = facets?.filter((facet) => facet.category === 'target_product') || [];
    for (let index = 0; index < productFacets.length; index++) {
      const productFacet = productFacets[index];
      const subProducts =
        productFacet.sub_facets?.filter((facet) => facet.category === 'sub_product').map(getDisplayName) || [];
      if (subProducts.length) {
        res = res.concat(subProducts);
      } else {
        res.push(getDisplayName(productFacet));
      }
    }

    return res;
  }

  const techArticleProps = {
    mainEntity: getTargetProductsNames(facets).map((name) => ({
      '@type': 'SoftwareApplication',
      name: StructuredData.addCompanyToName(name),
      applicationCategory: 'DeveloperApplication',
      offers: {
        price: 0,
        priceCurrency: 'USD',
      },
    })),
    headline: pageTitle,
  };

  const genres = getGenreNames(facets);
  if (genres.length) {
    techArticleProps['genre'] = genres;
  }

  return new TechArticleSd(techArticleProps);
};

export const constructHowToSd = ({ steps, parentHeading }) => {
  function getHowToSection(procedureDirective, name) {
    const howToSection = {
      '@type': 'HowToSection',
      name,
      itemListElement: [],
    };

    for (const step of procedureDirective.children) {
      handleStep(step, howToSection['itemListElement']);
    }

    return howToSection;
  }

  /**
   *
   * @param {node}    step
   * @param {step[]}  targetList  can be either steps[] of HowTo or itemListElement[] of HowToSection
   */
  function handleStep(step, targetList) {
    if (step['name'] !== 'step') {
      return;
    }
    // text of step is derived from children, or fallback to step's argument
    const childText = getPlaintext(step.children);
    const argText = getPlaintext(step.argument);
    // NOTE: step.argument is repeated in step.children as a Heading component
    // so strip the heading from children
    const bodyText = childText.replace(argText, '');

    // deep search for nested procedure to make sibling sections
    const nestedProcedure = findKeyValuePair(step.children, 'name', 'procedure');
    if (nestedProcedure) {
      targetList.push(getHowToSection(nestedProcedure, argText || bodyText));
    } else {
      // build step
      const stepSD = {
        '@type': 'HowToStep',
        text: bodyText.length ? bodyText : argText,
      };
      if (bodyText.length && argText) {
        stepSD['name'] = argText;
      }
      targetList.push(stepSD);
    }
  }

  const howToProps = {
    steps: [],
    name: parentHeading,
  };

  for (const step of steps) {
    handleStep(step, howToProps['steps']);
  }

  return new HowToSd(howToProps);
};
