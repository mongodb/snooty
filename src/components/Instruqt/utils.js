export const getLabURL = (embedValue) => {
  return `https://play.instruqt.com/embed${embedValue}`;
};
