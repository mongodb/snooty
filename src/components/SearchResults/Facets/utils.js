/**
 *
 * @param {obj} facet - contains key and id properties as returned by /v2/status ie.
 * https://docs-search-transport.mongodb.com/v2/status
 *
 * @returns {str} blue | green
 */
export const getFacetTagVariant = (facet) => {
  if (facet.key === 'target_product') {
    return 'green';
  }
  return 'blue';
};
