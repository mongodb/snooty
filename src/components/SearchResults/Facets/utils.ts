/**
 *
 * @param {obj} facet - contains key and id properties as returned by /v2/status ie.
 * https://docs-search-transport.mongodb.com/v2/status
 */
export const getFacetTagVariant = (facet: { key: string; id: string }) => {
  if (facet.key === 'target_product') {
    return 'green';
  }
  return 'blue';
};
