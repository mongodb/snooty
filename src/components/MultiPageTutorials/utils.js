import { reportAnalytics } from '../../utils/report-analytics';

export const reportMPTAnalytics = (targetSlug, variant) => {
  reportAnalytics('MultiPageTutorialNextClicked', {
    targetSlug,
    variant,
  });
};
