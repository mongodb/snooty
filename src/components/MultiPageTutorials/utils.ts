import { reportAnalytics } from '../../utils/report-analytics';

export const reportMPTAnalytics = (targetSlug: string, variant: string) => {
  reportAnalytics('MultiPageTutorialNextClicked', {
    targetSlug,
    variant,
  });
};
