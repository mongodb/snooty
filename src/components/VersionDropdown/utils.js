import { generatePathPrefix } from '../../utils/generate-path-prefix';

/**
 * Generates the prefix to be used for a version's URL. The prefix will typically consist of the docs repo's
 * set prefix, with the new version appended at the end. Different environments may return different
 * prefixes because of what's passed into the frontend.
 * @param {string} version The version to include at the end of the prefix.
 * @param {Object} siteMetadata Metadata about the site, including its pathPrefix, project, and snootyEnv.
 * @param {string} siteBasePrefix The current docs site's base prefix to append the version to.
 */
export const generatePrefix = (version, siteMetadata, siteBasePrefix) => {
  const { pathPrefix, snootyEnv } = siteMetadata;

  // pathPrefix is typically passed in for deploy jobs. For development purposes,
  // the returned prefix should follow the same pattern (even though the resulting URL might 404).
  if (!!pathPrefix || snootyEnv === 'development') {
    return `/${siteBasePrefix}/${version}`;
  }

  // For staging, replace current version in dynamically generated path prefix
  return generatePathPrefix({ ...siteMetadata, parserBranch: version });
};
