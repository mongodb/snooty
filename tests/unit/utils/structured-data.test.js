import { SoftwareSourceCodeSd, VideoObjectSd, constructHowToSd } from '../../../src/utils/structured-data';
import stepsData from '../../utils/data/how-to-structured-data.json';

describe('Structured Data', () => {
  describe('HowTo Structured Data', () => {
    it('converts steps into expected structured data format', () => {
      const howToSd = constructHowToSd({ steps: stepsData });
      expect(howToSd).toMatchSnapshot();
    });
  });

  describe('SoftwareSourceCode', () => {
    it('returns valid structured data with programmingLanguage', () => {
      const code = 'print("hello world")';
      const lang = 'py';
      const softwareSourceCodeSd = new SoftwareSourceCodeSd({ code, lang });
      expect(softwareSourceCodeSd.isValid()).toBeTruthy();
      expect(softwareSourceCodeSd).toMatchSnapshot();
    });

    it('returns valid structured data without programmingLangauge', () => {
      const code = 'print("hello world")';
      const softwareSourceCodeSd = new SoftwareSourceCodeSd({ code });
      expect(softwareSourceCodeSd.isValid()).toBeTruthy();
      expect(softwareSourceCodeSd).toMatchSnapshot();
    });
  });

  describe('VideoObject', () => {
    const embedUrl = 'https://www.youtube.com/embed/XrJG994YxD8';
    const name = 'Mastering Indexing for Perfect Query Matching';
    const uploadDate = '2023-11-08T05:00:28-08:00';
    const thumbnailUrl = 'https://i.ytimg.com/vi/XrJG994YxD8/maxresdefault.jpg';
    const description = 'Learn more about indexes in Atlas Search';

    it('returns valid structured data with description', () => {
      const videoObjectSd = new VideoObjectSd({ embedUrl, name, uploadDate, thumbnailUrl, description });
      expect(videoObjectSd.isValid()).toBeTruthy();
      expect(videoObjectSd).toMatchSnapshot();
    });

    it('returns valid structured data without description', () => {
      const videoObjectSd = new VideoObjectSd({ embedUrl, name, uploadDate, thumbnailUrl });
      expect(videoObjectSd.isValid()).toBeTruthy();
      expect(videoObjectSd).toMatchSnapshot();
    });

    it('returns invalid structured data with missing name field', () => {
      const videoObjectSd = new VideoObjectSd({ embedUrl, uploadDate, thumbnailUrl, description });
      expect(videoObjectSd.isValid()).toBeFalsy();
    });
  });
});
