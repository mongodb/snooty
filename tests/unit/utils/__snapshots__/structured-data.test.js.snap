// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Structured Data HowTo Structured Data converts steps into expected structured data format 1`] = `
HowToSd {
  "@context": "https://schema.org",
  "@type": "HowTo",
  "image": "https://webimages.mongodb.com/_com_assets/cms/kuyj2focmkbxv7gh3-stacked_default_slate_blue.svg?auto=format%252Ccompress",
  "name": undefined,
  "steps": [
    {
      "@type": "HowToStep",
      "name": "In Atlas, go to your federated database instance for your project.",
      "text": "If it's not already displayed, select the
organization that contains your project from the
 Organizations menu in the navigation bar.If it's not already displayed, select your project
from the Projects menu in the navigation bar.In the sidebar, click Data Federation under
the Services heading.The Data Federation page displays.",
    },
    {
      "@type": "HowToStep",
      "name": "Click Create Federated Database Instance.",
      "text": "If you have an existing federated database instance, instead click
Create Federated Database in the
top right corner of the dashboard.",
    },
    {
      "@type": "HowToSection",
      "itemListElement": [
        {
          "@type": "HowToStep",
          "name": "Rename the default collection.",
          "text": "Click  next to the default collection
VirtualCollection0 to edit its name. For this tutorial,
rename your collection Sessions.",
        },
        {
          "@type": "HowToStep",
          "name": "Create a second collection.",
          "text": "Click  next to the default name
VirtualDatabase0 to add a collection to the database.
For this tutorial, name your new collection Users.",
        },
        {
          "@type": "HowToStep",
          "name": "Add data to your virtual database.",
          "text": "Drag and drop the following data sources into the
respective federated database instance virtual collections:/mflix/sessions.json, into the Sessions
collection, and/mflix/users.json into the Users collection.",
        },
      ],
      "name": "Connect to a Data Source and add sample data to your federated database instance.",
    },
    {
      "@type": "HowToStep",
      "name": "Click Save.",
      "text": "Your federated database instance appears on the Data Federation page.",
    },
  ],
}
`;

exports[`Structured Data SoftwareSourceCode returns valid structured data with programmingLanguage 1`] = `
SoftwareSourceCodeSd {
  "@context": "https://schema.org",
  "@type": "SoftwareSourceCode",
  "codeSampleType": "code snippet",
  "text": "print("hello world")",
}
`;

exports[`Structured Data SoftwareSourceCode returns valid structured data without programmingLangauge 1`] = `
SoftwareSourceCodeSd {
  "@context": "https://schema.org",
  "@type": "SoftwareSourceCode",
  "codeSampleType": "code snippet",
  "text": "print("hello world")",
}
`;

exports[`Structured Data VideoObject returns valid structured data with description 1`] = `
VideoObjectSd {
  "@context": "https://schema.org",
  "@type": "VideoObject",
  "description": "Learn more about indexes in Atlas Search",
  "embedUrl": "https://www.youtube.com/embed/XrJG994YxD8",
  "name": "Mastering Indexing for Perfect Query Matching",
  "thumbnailUrl": "https://i.ytimg.com/vi/XrJG994YxD8/maxresdefault.jpg",
  "uploadDate": "2023-11-08T05:00:28-08:00",
}
`;

exports[`Structured Data VideoObject returns valid structured data without description 1`] = `
VideoObjectSd {
  "@context": "https://schema.org",
  "@type": "VideoObject",
  "embedUrl": "https://www.youtube.com/embed/XrJG994YxD8",
  "name": "Mastering Indexing for Perfect Query Matching",
  "thumbnailUrl": "https://i.ytimg.com/vi/XrJG994YxD8/maxresdefault.jpg",
  "uploadDate": "2023-11-08T05:00:28-08:00",
}
`;
